components/sync15/src/client/sync.rs

```diff
@@ -40,54 +40,72 @@ pub fn synchronize_with_clients_engine(
-        log::info!("skipping incoming for {} - not needed.", collection);
-        vec![IncomingChangeset::new(collection, coll_state.last_modified)]
-    } else {
-        assert_eq!(collection_requests.last().unwrap().collection, collection);
-
-        let count = collection_requests.len();
-        collection_requests
-            .into_iter()
-            .enumerate()
-            .map(|(idx, collection_request)| {
-                interruptee.err_if_interrupted()?;
-                let incoming_changes =
-                    super::fetch_incoming(client, &coll_state, collection_request)?;
-
-                log::info!(
-                    "Downloaded {} remote changes (request {} of {})",
-                    incoming_changes.changes.len(),
-                    idx,
-                    count,
-                );
-                Ok(incoming_changes)
-            })
-            .collect::<Result<Vec<_>, Error>>()?
+            // Doing this batching will involve specifying a "limit=" param and
+            // "x-if-unmodified-since" for each request, looking for an
+            // "X-Weave-Next-Offset header in the response and using that in subsequent
+            // requests.
+            // See https://mozilla-services.readthedocs.io/en/latest/storage/apis-1.5.html#syncstorage-paging
+            //
+            // But even if we had that, we need to deal with a 412 response on a subsequent batch,
+            // so we can't know if we've staged *every* record for that timestamp; the next
+            // sync must use an earlier one.
+            //
+            // For this reason, an engine can't really trust a server timestamp until the
+            // very end when we know we've staged them all.
+            let incoming = super::fetch_incoming(client, &coll_state, collection_request)?;
+            log::info!("Downloaded {} remote changes", incoming.len());
+            engine.stage_incoming(incoming, telem_engine)?;
+            interruptee.err_if_interrupted()?;
+        }
     };
 
-    let outgoing = engine.apply_incoming(incoming, telem_engine)?;
+    // Should probably add a new `fetch_outgoing()` and only call `apply()` when something was
+    // staged?
+    log::info!("Applying changes");
+    let outgoing = engine.apply(coll_state.last_modified, telem_engine)?;
     interruptee.err_if_interrupted()?;
-    log::info!("Uploading {} outgoing changes", outgoing.changes.len());
-    let upload_info =
-        CollectionUpdate::new_from_changeset(client, &coll_state, outgoing, fully_atomic)?
-            .upload()?;
+
+    // XXX - this upload strategy is buggy due to batching. With enough records, we will commit
+    // 2 batches on the server. If the second fails, we get an Err<> here, so can't tell the
+    // engine about the successful server batch commit.
+    // Most stuff below should be called per-batch rather than at the successful end of all
+    // batches, but that's not trivial.
```

Comment:
Ooof...yeah, I think the lack of per-batch calls (and reading everything in one request) is an artifact of us implementing batches "transparently" on Desktop way back when, and carrying that through all our implementations.

In case of the first batch failing and the second succeeding, I think we decided that not advancing the last sync time after the successful batch, and reading our own changes on the next sync, was good enough back then...but it would sure be nice to handle them properly.

Reply:
yeah, exactly - I think the new shape of the trait is now capable of expressing this correctly, so it's just a matter of having sync15 actually do it.